File: android/app/src/main/java/com/pearsonmedia/lociate/domain/model/Locus.kt
package com.pearsonmedia.loci.domain.model

import java.time.Instant
import java.util.UUID

data class Locus(
    val id: UUID = UUID.randomUUID(),
    val latitude: Double,
    val longitude: Double,
    val locationName: String? = null,
    val audioFilePath: String,
    val transcription: String = "",
    val category: LocusCategory = LocusCategory.GENERAL,
    val isShared: Boolean = false,
    val createdByName: String? = null,
    val householdId: UUID? = null,
    val createdAt: Instant = Instant.now(),
    val updatedAt: Instant = Instant.now(),
    val isArchived: Boolean = false,
    val syncStatus: SyncStatus = SyncStatus.PENDING
)

enum class SyncStatus {
    PENDING,
    SYNCED,
    MODIFIED,
    CONFLICTED
}

enum class SubscriptionTier {
    FREE,
    PREMIUM,
    FAMILY;

    val maxLoci: Int
        get() = when (this) {
            FREE -> 10
            PREMIUM, FAMILY -> Int.MAX_VALUE
        }

    val hasCloudSync: Boolean
        get() = this != FREE

    val hasAiCategorization: Boolean
        get() = this != FREE

    val hasWidget: Boolean
        get() = this != FREE

    val hasSearch: Boolean
        get() = this != FREE

    val hasFamilySharing: Boolean
        get() = this == FAMILY

    val maxHouseholdMembers: Int
        get() = if (this == FAMILY) 6 else 0
}

File: Lociate/Models/Locus.swift
import CoreLocation
import Foundation
import SwiftData

@Model
final class Locus {
    @Attribute(.unique) var id: UUID
    var latitude: Double
    var longitude: Double
    var locationName: String?
    var audioFileURL: String
    var transcription: String
    var categoryRawValue: String
    var isShared: Bool
    var createdByName: String?
    var householdId: UUID?
    var createdAt: Date
    var updatedAt: Date
    var isArchived: Bool
    var syncStatusRawValue: String

    var category: LocusCategory {
        get { LocusCategory(rawValue: categoryRawValue) ?? .general }
        set { categoryRawValue = newValue.rawValue }
    }

    var coordinate: CLLocationCoordinate2D {
        CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
    }

    var syncStatus: SyncStatus {
        get { SyncStatus(rawValue: syncStatusRawValue) ?? .pending }
        set { syncStatusRawValue = newValue.rawValue }
    }

    init(
        id: UUID = UUID(),
        latitude: Double,
        longitude: Double,
        locationName: String? = nil,
        audioFileURL: String,
        transcription: String = "",
        category: LocusCategory = .general,
        isShared: Bool = false,
        createdByName: String? = nil,
        householdId: UUID? = nil,
        createdAt: Date = Date(),
        updatedAt: Date = Date(),
        isArchived: Bool = false,
        syncStatus: SyncStatus = .pending
    ) {
        self.id = id
        self.latitude = latitude
        self.longitude = longitude
        self.locationName = locationName
        self.audioFileURL = audioFileURL
        self.transcription = transcription
        self.categoryRawValue = category.rawValue
        self.isShared = isShared
        self.createdByName = createdByName
        self.householdId = householdId
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.isArchived = isArchived
        self.syncStatusRawValue = syncStatus.rawValue
    }
}

enum SyncStatus: String, Codable {
    case pending
    case synced
    case modified
    case conflicted
}
